File: COURSE 1 (Algorithmic Toolbox)/week3_greedy_algorithms/0.Week 3 Assignment/submitted files c++/car_fueling.cpp
#include <iostream>
#include <vector>
using namespace std;
using std::cin;
using std::cout;
using std::vector;
using std::max;

int compute_min_refills(int dist, int tank, vector<int> & stops) {

        int numofrefills=0;
        int currentrefill=0;

        vector<int> allstops(stops.size()+2);

        allstops[0]=0;
        allstops[allstops.size()-1]=dist;

        for(int i=1;i<allstops.size()-1;i++){
            allstops[i]=stops[i-1];
        }


        while (currentrefill<allstops.size()-1){
            int lastrefill=currentrefill;

            while (currentrefill<allstops.size()-1 && allstops[currentrefill+1]-allstops[lastrefill]<=tank){
                currentrefill+=1;
            }

            if(currentrefill==lastrefill){
                return -1;
            }

            if(currentrefill<=allstops.size()-1){
                numofrefills+=1;
            }
        }
        return numofrefills-1;
}


int main() {
    int d = 0;
    cin >> d;
    int m = 0;
    cin >> m;
    int n = 0;
    cin >> n;

    vector<int> stops(n);
    for (size_t i = 0; i < n; ++i)
        cin >> stops.at(i);

    cout << compute_min_refills(d, m, stops) << "\n";

    return 0;
}

File: COURSE 1 (Algorithmic Toolbox)/week3_greedy_algorithms/0.Week 3 Assignment/submitted files java/CarFueling.java
import java.util.*;
import java.io.*;

public class CarFueling {
    static int computeMinRefills(int dist, int tank, int[] stops) {

        int numofrefills=0;
        int currentrefill=0;

        int[] allstops=new int[stops.length+2];

        allstops[0]=0;
        allstops[allstops.length-1]=dist;

        for(int i=1;i<allstops.length-1;i++){
            allstops[i]=stops[i-1];
        }


        while (currentrefill<allstops.length-1){
            int lastrefill=currentrefill;

            while (currentrefill<allstops.length-1 && allstops[currentrefill+1]-allstops[lastrefill]<=tank){
                currentrefill+=1;
            }

            if(currentrefill==lastrefill){
                return -1;
            }

            if(currentrefill<=allstops.length-1){
                numofrefills+=1;
            }
        }
        return numofrefills-1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int dist = scanner.nextInt();
        int tank = scanner.nextInt();
        int n = scanner.nextInt();
        int[] stops = new int[n];
        for (int i = 0; i < n; i++) {
            stops[i] = scanner.nextInt();
        }

        System.out.println(computeMinRefills(dist, tank, stops));
    }
}
